### PR TITLE
Add `TF_STATE_PERSIST_INTERVAL` environment variable. 

### LINKED ISSUE #35115: Make persistence interval configurable for the remote state backend
### Terraform Version

```shell
1.8.1
```

### Use Cases

When state size increases, the persistence of the state becomes a bottleneck as it happens synchronously. Currently, the persist interval is [hardcoded in local apply backend](https://github.com/hashicorp/terraform/blob/main/internal/backend/local/backend_apply.go#L85) to 20 seconds, and as state size increases, it becomes a significant amount of time spent on the persisting. 

For remote apply backend it's possible to disable intermediate persistence completely, but it's not possible to do for the local apply backend.

### Attempted Solutions

I have code changes that allow to specify a new persist interval via an environment variable.

### Proposal

Make the persist interval configurable (or allow you to disable it completely) instead of using a hardcoded value. Open question: How to specify it: Is it enough to provide support via an environment variable, or will we need to add support for the command-line option?

Also, maybe it makes sense to make actual cloud write async? 

### References

_No response_

### PR DESCRIPTION
**Pull request by sebasslash**
This PR stems from the initial work done by @alexott (#35116), which fixes #35115. These changes are a simple refactor to delegate handling this env var to `command.Meta`. This allows us to pass the user specified persist interval to `backendrun.Operation` and keep env var handling (with existing vars such as `TF_WORKSPACE`, `TF_INPUT`) centralized within `command.Meta`.

The default interval remains 20 seconds and serves as a global minimum interval. Remote backends such as `cloud` may override the default using the `X-Terraform-Snapshot-Interval` header. If the default interval value is greater than the user-specified value, the default will be used. 

## Target Release

1.9.x

### ENHANCEMENTS

 

-  Added TF_STATE_PERSIST_INTERVAL environment variable to control persist interval in the remote state backend.

### CODE REVIEW

**Comment by sebasslash** [comment]
Still missing updating the docs and adding a few more tests, if possible.

File: internal/command/meta.go [hunk 1]
```diff
@@ -736,6 +736,27 @@ func (m *Meta) showDiagnostics(vals ...interface{}) {
 	}
 }
 
+const (
+	// StatePersistIntervalEnvVar is the environment variable that can be set
+	// to control the interval at which Terraform persists state. The interval
+	// itself defaults to 20 seconds.
+	StatePersistIntervalEnvVar = "TF_STATE_PERSIST_INTERVAL"
```
Comment (brandonc) [inline internal/command/meta.go:743]: What if this mirrored the header configuration to call out the feature in question? TF_STATE_SNAPSHOT_INTERVAL

File: internal/command/meta.go [hunk 2]
```diff
@@ -736,6 +736,27 @@ func (m *Meta) showDiagnostics(vals ...interface{}) {
...
+// StatePersistInterval returns the configured interval that Terraform should
+// persist statefiles to the desired backend. Backends may choose to override
+// the default value.
+func (m *Meta) StatePersistInterval() int {
+	if val, ok := os.LookupEnv(StatePersistIntervalEnvVar); ok {
+		interval, err := strconv.Atoi(val)
+		if err == nil {
+			return interval
+		}
+		log.Printf("[ERROR] Can't parse persist interval '%s' of environment variable '%s'", val, StatePersistIntervalEnvVar)
```
Comment (brandonc) [inline internal/command/meta.go:755]: 💅 standard quote formatting

```suggestion
		log.Printf("[ERROR] Can't parse persist interval %q of environment variable %q", val, StatePersistIntervalEnvVar)
```

**Review by brandonc — APPROVED** [review]
Great, thank you.

File: internal/command/meta.go (hunk 1 above)
Comment (sebasslash) [reply internal/command/meta.go:743]: I can't remember the exact conversation around this 😅, but I was under the impression we didn't want to use the term "snapshot" in a public context (even though the HTTP header uses the term).

File: internal/command/meta.go [hunk 3]
```diff
@@ -736,6 +736,27 @@ func (m *Meta) showDiagnostics(vals ...interface{}) {
...
+)
+
+// StatePersistInterval returns the configured interval that Terraform should
+// persist statefiles to the desired backend. Backends may choose to override
+// the default value.
+func (m *Meta) StatePersistInterval() int {
+	if val, ok := os.LookupEnv(StatePersistIntervalEnvVar); ok {
+		interval, err := strconv.Atoi(val)
+		if err == nil {
+			return interval
```
Comment (alexott) [inline internal/command/meta.go:753]: Should we add guards here - like, don't allow to specify negative values, and maybe put 20 as a minimum value? We really need only much higher values, like 600, 1200, etc.

**Review by alexott — COMMENTED** [review]
thank you - just one comment about boundary check

**Review by brandonc — APPROVED** [review]

File: website/docs/cli/config/environment-variables.mdx [hunk 4]
```diff
@@ -143,6 +143,14 @@ The default client timeout for requests to the remote registry is 10s. `TF_REGIS
 export TF_REGISTRY_CLIENT_TIMEOUT=15
 ```
 
+## TF_STATE_PERSIST_INTERVAL
+
+The interval in seconds Terraform will attempt to persist state to a remote backend during an apply operation. The default minimum interval for all remote backends is 20 seconds, however certain backends may choose to override the default value. If the value of this interval is lower than the default interval, the default will be used.
```
Comment (rkoron007) [inline website/docs/cli/config/environment-variables.mdx:148]: Small style suggestions for active voice and present tense:
```suggestion
The interval in seconds that Terraform attempts to persist state to a remote backend during an apply operation. The default minimum interval for all remote backends is 20 seconds. Backends may override the default value, but Terraform uses the default if a backend's specified interval is less than the default interval.
```

**Pull request merged**